numpy/core/src/multiarray/datetime.c

```diff
@@ -462,8 +462,21 @@ convert_datetime_to_datetimestruct(PyArray_DatetimeMetaData *meta,
             break;
 
         case NPY_FR_W:
-            /* A week is 7 days */
-            set_datetimestruct_days(dt * 7, out);
+            /*
+             * A week is 7 days but we can overflow
+             * if we apply the product directly, so
+             * iteratively account for week when near
+             * max int64 value
+             */
+            if (dt > 0 && (dt > (NPY_MAX_INT64 / 7)) ||
+                (dt < 0 && (dt < NPY_MIN_INT64 / 7))) {
+                for (int i = 0; i < 7; ++i) {
+                    set_datetimestruct_days(dt, out);
+                }
```

Comment:
I wonder if this awkward or elegant--a week naturally breaks down into 7 days, but I also thought there may be a more performant solution using i.e., modulus to get the true number of loops needed to avoid overflow.

Reply:
I think this is fine. It's better to be overly conservative and correct in some edge cases than to be maximally performant :)